### PR TITLE
Lwt awareness

### LINKED ISSUE #24: Prefer primary replica for the token for LWT queries
The way Paxos protocol works is that if two queries attempt to update the same key from different coordinators, they will start two independent Paxos rounds. Each round will be assigned a timestamp, and the coordinator who has the highest timestamp will win.
The issue is, the loser only queues up after the winner (using a semaphore) if both rounds are coordinated by the same node. If rounds are started at different nodes, the only option for the user is to sleep increasingly random interval and retry (this is what our implementation does).

If the key is contended and the driver is neither shard nor token aware, this leads to a lot of retries to make an update. 
If the key is contended and the driver is token and shard aware, it will send the query to one of the replicas for the partition, but will round-robin between replicas.

This still means there will be at least 50% of loser queries, which will retry before they can commit. Retries against a contended key in multi-DC setup lead to increasingly growing delays and time outs. If it takes 100 milliseconds to do a round due to network topology, and the key is contended, we get as few as 1 query per second for such a key. If all queries queue up on the same coordinator, we can get to up to 10 QPS per key.

This is why for LWT queries the driver should choose replicas in a pre-defined order, so that in case of contention they will queue up at the same replica, rather than compete: choose the primary replica first, then, if the primary is known to be down, the first secondary, then the second secondary, and so on.

This will reduce contention over hot keys and thus increase LWT performance.

If LOCAL_SERIAL serial consistency is used, we should prefer the primary in the local DC, because only local DCs endpoints will participate in the Paxos round. If SERIAL consistency is specified in a multi-DC setup, we should use any DCs primary, but consistently use the same primary for all queries on all clients. The key to avoiding contention is in all clients consistently choosing the same replica for the same key, if it's available/alive.

See also https://github.com/scylladb/gocql/issues/40

### PR DESCRIPTION
**Pull request by Lorak-mmk**
PR based on https://github.com/scylladb/java-driver/pull/122, so should get merged after it.

Fixes #24 

PR is generally a port of commits mentioned in above issue, but modified to accommodate different architecture of driver 4.x, and also includes some integration tests.

### CODE REVIEW

File: integration-tests/src/test/java/com/datastax/oss/driver/core/cql/PreparedStatementIT.java [hunk 1]
```diff
@@ -474,6 +475,34 @@ public void should_infer_routing_information_when_partition_key_is_bound() {
         "DELETE FROM prepared_statement_test WHERE a = ?");
   }
 
+  @Test
+  @CassandraSkip // Functionality only available in Scylla
```
Comment (avelanarius) [inline integration-tests/src/test/java/com/datastax/oss/driver/core/cql/PreparedStatementIT.java:479]: You said that you saw that `@CassandraSkip` was not honored. I saw something like this when I fixing original integration tests and I worked around it this way: https://github.com/scylladb/java-driver/blob/7381fb9f8c5a27bee994194e686b60178977940e/integration-tests/src/test/java/com/datastax/oss/driver/core/cql/reactive/DefaultReactiveResultSetIT.java#L180

The original DataStax tests also did it that way, so I didn't spend any time debugging this, but thinking more about this I think you should try to debug this problem. What's strange is there are other annotations `@ScyllaSkip` in this file and they are correctly honored.

**Comment by avelanarius** [comment]
@Lorak-mmk Please also look into #127 - we have one customer that wants this functionality.

File: integration-tests/src/test/java/com/datastax/oss/driver/core/cql/PreparedStatementIT.java (hunk 1 above)
Comment (Lorak-mmk) [reply integration-tests/src/test/java/com/datastax/oss/driver/core/cql/PreparedStatementIT.java:479]: I found the issue. The adnotation is honored in `PreparedStatementIT`, but not in `LWTLoadBalancingIT` - in the latter it is only checked at class-level, not method-level. The difference seems to be `@Category(ParallelizableTests.class)`. Unfortunately, my new test class can't use this adnotation - for the test to be parallelizable it must use exactly 1 node, my tests have to use more. So I'll use the ` assumeThat(CcmBridge.SCYLLA_ENABLEMENT).isFalse();` solution (or rather: `assumeTrue(CcmBridge.SCYLLA_ENABLEMENT);`).

**Comment by Lorak-mmk** [comment]
Now the tests are passing.

**Comment by Lorak-mmk** [comment]
#122 is now merged, so I rebased on scylla-4.x

**Review by avelanarius — APPROVED** [review]
LGTM. When the integration tests finish running locally I will merge it.

**Pull request merged**